smart-contracts/contracts/Unlock.sol

```diff
@@ -509,6 +527,116 @@ contract Unlock is UnlockInitializable, UnlockOwnable {
     }
   }
 
+  /**
+   * Swap tokens and call a function in a lock contract
+   */
+  function swapAndCall(
+    address lock,
+    address srcToken,
+    uint amountInMax,
+    uint keyPrice,
+    uint24 poolFee,
+    bytes memory callData
+  ) public payable returns(bytes memory) {
+    // make sure uniswap is set
+    if(address(uniswapRouter) == address(0)) {
+      revert UniswapNotSet();
+    }
+
+    // get lock pricing 
+    address destToken = IPublicLock(lock).tokenAddress();
+    if (srcToken == destToken) {
+      revert('=');
+    }
+    
+    // TODO: fetch automatically price from lock
+    // uint keyPrice = IPublicLock(lock).keyPrice();
+    // make sure amount is enough to unwrap
+    // if(srcToken == address(0) && amountInMax < msg.value) {
+    //   revert InsufficientAmount(srcToken);
+    // }
+
+    // use WETH for native tokens
+    address tokenIn = srcToken == address(0) ? weth : srcToken;
+
+
+    // wrap native tokens
+    if(srcToken == address(0)) {      
+      IWETH(weth).deposit{value: amountInMax}();
+    } else {
+      // Transfer the specified amount of ERC20 to this contract
+      TransferHelper.safeTransferFrom(srcToken, msg.sender, address(this), amountInMax);
```

Comment:
If we do this, let's make sure the `lock` is actually a lock (I think we have a helpfer function for that already)

Reply:
I want to avoid re-entrancy and/or bad contracts!

Reply:
I also want to test the flow where a proper/legit lock has a malicious hook which could be use to "steal" the funds from Unlock?